examples/sir_hmc.py

```diff
@@ -0,0 +1,632 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import logging
+import math
+import re
+
+import torch
+from torch.autograd import grad
+from torch.distributions import biject_to, constraints
+from torch.distributions.transforms import ComposeTransform
+
+import pyro
+import pyro.distributions as dist
+import pyro.distributions.hmm
+import pyro.poutine as poutine
+from pyro.distributions.transforms.discrete_cosine import DiscreteCosineTransform
+from pyro.infer import MCMC, NUTS, config_enumerate, infer_discrete
+from pyro.infer.autoguide import init_to_value
+from pyro.infer.mcmc.util import TraceEinsumEvaluator
+from pyro.infer.reparam import DiscreteCosineReparam
+from pyro.ops.tensor_utils import convolve, safe_log
+from pyro.util import warn_if_nan
+
+logging.basicConfig(format='%(message)s', level=logging.INFO)
+
+
+# Introduction
+# ============
+#
+# First consider a simple SIR model (Susceptible Infected Recovered).
+#
+# Note we need to use ExtendedBinomial rather than Binomial because the data
+# may lie outside of the predicted support. For these values,
+# Binomial.log_prob() will error, whereas ExtendedBinomial.log_prob() will
+# return -inf.
+
+def global_model(population):
+    R0 = pyro.sample("R0", dist.LogNormal(0., 1.))
+    tau = pyro.sample("tau", dist.LogNormal(math.log(7.), 1.))
+    rho = pyro.sample("rho", dist.Uniform(0, 1))
+
+    rate_s = -R0 / (tau * population)
+    prob_i = 1 / (1 + tau)
+
+    return rate_s, prob_i, rho
+
+
+def discrete_model(data, population):
+    # Sample global parameters.
+    rate_s, prob_i, rho = global_model(population)
+
+    # Sequentially filter.
+    S = torch.tensor(population - 1.)
+    I = torch.tensor(1.)
+    for t, datum in enumerate(data):
+        S2I = pyro.sample("S2I_{}".format(t),
+                          dist.Binomial(S, -(rate_s * I).expm1()))
```

Comment:
is the `I` supposed to be part of the `expm1`?

Reply:
Yes, the formula is `1 - exp(-blah)` where `blah = R0 / tau * I / population`.